### PR TITLE
Added main code for the multitask and DA experiments.

### PR DESCRIPTION
**Pull request by dbash**

### CODE REVIEW

File: lib/multitask.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
+
```
Comment (azhmogin-google) [inline lib/multitask.py:3]: Put a file docstring ("""...""") here.

File: lib/multitask.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+  parser.add_argument("--finetune_mode",
+                      choices=["h", "m", "b", "hm", "hb", "hbm", "all"],
+                      default="h",
+                      help="finetuning mode "
+                      "(h = head, m = mixture weights,"
+                      "b = batch norm, all = entire model.)")
+  return parser
+
+def get_losses(datasets, num_train_ds, label_smoothing=0):
+  """Returens dictionaries of losses and loss weights for the datasets.
```
Comment (azhmogin-google) [inline lib/multitask.py]:
```suggestion
  """Returns dictionaries of losses and loss weights for the datasets.
```

File: lib/multitask.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+  return losses, loss_weights
+
+def get_dataset_list(dataset_type, num_datasets=None):
+  """Maps the dataset type with the list of dataset names.
+
+  Arguments:
+  dataset_type: datasets type.
+  num_datasets: optional number of datasets to use.
+  """
+  switcher = {
```
Comment (azhmogin-google) [inline lib/multitask.py]: I would move this outside of the function into the global scope (with a more specific name too).

File: lib/multitask.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+  Arguments:
+  datasets_dict: dictionary containing the datasets (to be acquired with
+  `get_datasets` method).
+  """
+  train_x_datasets = [datasets_dict[x]["train_x"] for x in datasets_dict]
+  train_y_datasets = [datasets_dict[x]["train_y"] for x in datasets_dict]
+  test_x_datasets = [datasets_dict[x]["test_x"] for x in datasets_dict]
+  test_y_datasets = [datasets_dict[x]["test_y"] for x in datasets_dict]
+
+  combined_dataset_train_x = tf.data.Dataset.zip(tuple(train_x_datasets))
```
Comment (azhmogin-google) [inline lib/multitask.py:131]: Better restructure it along the lines:

```python
... (something above this too) ...
combined_dataset_train = [tf.data.Dataset.zip(...) for ds in [train_x_datasets, ...]]
combined_dateset_train = tf.data.Dataset.zip(tuple(combined_dataset_train))
```

File: lib/multitask.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+        out_adapter=out_adapter, kernel_regularizer=kernel_reg,
+        dropout=dropout, with_head=False, name="feature_extractor")
+    combined_model = build_models(
+        feature_extractor=feature_extractor, datasets=datasets_info,
+        input_shape=input_shape)
+  else:
+    feature_extractor = sresnet.shared_resnet(
+        input_shape=input_shape, num_layers=num_layers,
+        num_templates=num_templates, num_classes=10, tensor_size=tensor_size,
+        in_adapter=in_adapter, out_adapter=out_adapter,
```
Comment (azhmogin-google) [inline lib/multitask.py]: Many of these arguments can be shared (across `resnet` and `shared_resnet`) via a dictionary (with `**args` in the call).

File: lib/multitask.py [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+
+def fix_weights(model, target_dataset, finetune_mode="h", shared=False):
+  """Fixes model weights according to experiment type.
+
+  Arguments:
+  model: the tf.keras.Model object.
+  target_dataset: string target dataset name.
+  finetune_mode: a string combination of  the following characters that
+    tells which variables to set trainable.
+    Options:`h` - model head, `m` - mixture weights,
```
Comment (azhmogin-google) [inline lib/multitask.py:400]: A better approach is to use `enum.enum`

**Review by azhmogin-google — APPROVED** [review]

File: lib/multitask.py (hunk 6 above)
Comment (dbash) [reply lib/multitask.py:400]: The tricky part here is that the mode can consist of a combination of modes, e.g. "hm" would stand for setting the head and mixture weights as trainable. I have looked up the enum.enum documentation and the only suitable option would be an enum.Flag. The problem, though, is that we cannot set is as an argument in argParse and there is no obvious way to acquire an enum.Flag object from the string without ugly for loops through the string characters. Or maybe I don't understand something.

**Pull request merged**